Added support for Homebrew release

sabre1041 opened the pull request:
Added support for release to [Homebrew Tap](https://github.com/redhat-cop/homebrew-redhat-cop)

@bserdar Need to sync up with you to set repository secrets to enable pushing
@rpelisse Feel free to review

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+          HOMEBREW_TAP_REPO="${HOMEBREW_TAP_REPO:-redhat-cop/homebrew-redhat-cop}"
+          REPO_NAME=$(basename $HOMEBREW_TAP_REPO)
+
+          SHA256=`sha256sum target/jcliff-${JCLIFF_TAG:1}-dist.tar.gz | cut -f1 -d ' '`
```
@sabre1041 you are way too young to use `` instead of $() ;)

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+        run: |
+          set -e
+
+          JCLIFF_TAG=${{ steps.get_version.outputs.TAG_VERSION }}
```
I would prefix all those vars with readonly (or use lowercase)

rpelisse replied:
Also surround the var with double-quote

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+
+          JCLIFF_TAG=${{ steps.get_version.outputs.TAG_VERSION }}
+          HOMEBREW_TAP_REPO="${HOMEBREW_TAP_REPO:-redhat-cop/homebrew-redhat-cop}"
+          REPO_NAME=$(basename $HOMEBREW_TAP_REPO)
```
=> "${HOMEBREW_TAP_REPO}"

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+          SHA256=`sha256sum target/jcliff-${JCLIFF_TAG:1}-dist.tar.gz | cut -f1 -d ' '`
+          URL="https://github.com/bserdar/jcliff/releases/download/${JCLIFF_TAG}/jcliff-${JCLIFF_TAG:1}-dist.tar.gz"
+
+          rm -rf $REPO_NAME
```
=> "${REPO_NAME}"

especially here. Imagine:

REPO_NAME='a_file /' => boom root is erased!

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+
+          rm -rf $REPO_NAME
+
+          git clone https://github.com/${HOMEBREW_TAP_REPO}.git
```
I would add REPO_URL var and also add "" around the var substitution.

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+
+          git clone https://github.com/${HOMEBREW_TAP_REPO}.git
+
+          sed -i -e 's|^  url.*|  url "'"$URL"'"|g' $REPO_NAME/Formula/jcliff.rb
```
=> "${REPO_NAME}/Formula/jcliff.rb"

rpelisse replied:
=> "${URL}"

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+          git clone https://github.com/${HOMEBREW_TAP_REPO}.git
+
+          sed -i -e 's|^  url.*|  url "'"$URL"'"|g' $REPO_NAME/Formula/jcliff.rb
+          sed -i -e 's|^  sha256.*|  sha256 "'"$SHA256"'"|g' $REPO_NAME/Formula/jcliff.rb
```
=> "${SHA256}"

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+          sed -i -e 's|^  url.*|  url "'"$URL"'"|g' $REPO_NAME/Formula/jcliff.rb
+          sed -i -e 's|^  sha256.*|  sha256 "'"$SHA256"'"|g' $REPO_NAME/Formula/jcliff.rb
+
+          cd $(basename $HOMEBREW_TAP_REPO)
```
"${HOMEBREW_TAP_REPO}"

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+
+          cd $(basename $HOMEBREW_TAP_REPO)
+
+          git config user.name "Andrew Block"
```
Username, email => should be in vars.

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+          git add Formula/jcliff.rb
+          git commit -m "jcliff ${JCLIFF_TAG} release"
+
+          if [[ ! -z $BREW_USER && ! -z $BREW_TOKEN ]]; then
```
"${BREW_USER}" and "${BREW_TOKEN}"

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+          git commit -m "jcliff ${JCLIFF_TAG} release"
+
+          if [[ ! -z $BREW_USER && ! -z $BREW_TOKEN ]]; then
+            git push https://${BREW_USER}:${BREW_GITHUB_TOKEN}@github.com/${HOMEBREW_TAP_REPO}.git
```
=> add missing double quote : "https://${BREW_USER}:${BREW_GITHUB_TOKEN}@github.com/${HOMEBREW_TAP_REPO}.git"

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+          if [[ ! -z $BREW_USER && ! -z $BREW_TOKEN ]]; then
+            git push https://${BREW_USER}:${BREW_GITHUB_TOKEN}@github.com/${HOMEBREW_TAP_REPO}.git
+          else
+            echo "BREW_USER and BREW_TOKEN variables not set. Skipping push..."
```
echo "BREW_USER and BREW_TOKEN variables not set or are invalid:"
echo "BREW_USER: ${BREW_USER}"
echo "BREW_TOKEN: ${BREW_TOKEN}"
echo "Skipping push..."

rpelisse commented on .github/workflows/release.yml:
```diff
@@ -30,3 +30,38 @@ jobs:
+          HOMEBREW_TAP_REPO: ${{ secrets.HOMEBREW_REPOSITORY }}
+          BREW_USER: ${{ secrets.BREW_USER }}
+          BREW_TOKEN: ${{ secrets.BREW_TOKEN }}
+        run: |
```
@sabre1041 anyway to externalise this bash script into a separate file? It would be cleaner IMH.

sabre1041 replied:
Sure. Just need to know where in the repository it should go

rpelisse replied:
I generally put this kind of scripts into src/build/bash.

sabre1041 commented:
@rpelisse Addressed your concerns. The only area I did not address was for the Git URL. I dont see us going off of GitHub any time soon. However, we might want to look to migrating the Homebrew repository from the `redhat-cop` org to `ansible-middleware`.

bserdar commented:
@sabre1041 what do you need re: repository secrets?

rpelisse commented on src/main/scripts/homebrew-release.sh:
```diff
@@ -0,0 +1,30 @@
+git clone https://github.com/${HOMEBREW_TAP_REPO}.git
+
+sed -i -e 's|^  url.*|  url "'"$URL"'"|g' ${REPO_NAME}/Formula/jcliff.rb
+sed -i -e 's|^  sha256.*|  sha256 "'"$SHA256"'"|g' ${REPO_NAME}/Formula/jcliff.rb
```
It's probably overdoing it, but know that you can check the status of each commands in pipe using [the variable PIPESTATUS](https://tldp.org/LDP/abs/html/internalvariables.html#PIPESTATUSREF).

rpelisse reviewed (changes requested):
With all the missing overrides in place, we can further and have Bats unit test setup to validate that this script behave properly. (but it's probably overdoing it a bit)

rpelisse commented on src/main/scripts/homebrew-release.sh:
```diff
@@ -0,0 +1,30 @@
+set -e
+
+readonly JCLIFF_TAG="${{ steps.get_version.outputs.TAG_VERSION }}"
```
Of course... I really need to learn about Github Actions :) Please ignore.

rpelisse commented on src/main/scripts/homebrew-release.sh:
```diff
@@ -0,0 +1,30 @@
+
+cd $(basename "$HOMEBREW_TAP_REPO")
+
+git config user.name "${BREW_GITHUB_USER:-Andrew Block}"
```
@sabre1041 What about this?

rpelisse commented on src/main/scripts/homebrew-release.sh:
```diff
@@ -0,0 +1,30 @@
+sed -i -e 's|^  url.*|  url "'"$URL"'"|g' ${REPO_NAME}/Formula/jcliff.rb
+sed -i -e 's|^  sha256.*|  sha256 "'"$SHA256"'"|g' ${REPO_NAME}/Formula/jcliff.rb
+
+cd $(basename "$HOMEBREW_TAP_REPO")
```
@sabre1041 I think you missed this :)

rpelisse reviewed (approved):
There is a few things that can still be fixed up, but apart from those, I want to say good to go!

sabre1041 commented:
@rpelisse fixed. hopefully last one

rpelisse commented:
@sabre1041 yes, you have indulged my bash-extremism enough :) !
